hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -752,16 +770,85 @@ public Hashtable<String, String> getBlobMetadata(final Path path,
    */
   public void setBlobMetadata(final Path path,
       final Hashtable<String, String> metadata, TracingContext tracingContext)
-      throws AzureBlobFileSystemException {
+      throws IOException {
     try (AbfsPerfInfo perfInfo = startTracking("setBlobMetadata", "setBlobMetadata")) {
       LOG.debug("setBlobMetadata for filesystem: {} path: {} with properties: {}",
           client.getFileSystem(),
           path,
           metadata);
 
+      List<AbfsHttpHeader> metadataRequestHeaders = getRequestHeadersForMetadata(metadata);
+
+      try {
+        final AbfsRestOperation op = client.setBlobMetadata(path, metadataRequestHeaders, tracingContext);
+        perfInfo.registerResult(op.getResult()).registerSuccess(true);
+      } catch (AbfsRestOperationException ex) {
+        // The path does not exist explicitly.
+        // Check here if the path is an implicit dir
+        if (ex.getStatusCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+          List<BlobProperty> blobProperties = getListBlobs(
+              path,null, null, tracingContext, 2, true);
+          if (blobProperties.size() == 0) {
+            throw ex;
+          }
+          else {
+            // The path was an implicit blob. Create Marker and set metadata on it
+            createDirectory(path, null, FsPermission.getDirDefault(),
+                FsPermission.getUMask(
+                    getAbfsConfiguration().getRawConfiguration()),
+                tracingContext);
+
+            boolean xAttrExists = metadata.containsKey(HDI_ISFOLDER);
+            if (!xAttrExists) {
+              metadata.put(HDI_ISFOLDER, TRUE);
+              metadataRequestHeaders = getRequestHeadersForMetadata(metadata);
+            }
+
+            final AbfsRestOperation op = client.setBlobMetadata(path, metadataRequestHeaders, tracingContext);
+            perfInfo.registerResult(op.getResult()).registerSuccess(true);
+          }
+        }
+        else {
+          throw ex;
+        }
+      }
+    }
+  }
+
+  /**
+   * Gets user-defined properties(metadata) of the container over blob endpoint.
+   * @param tracingContext
+   * @return hashmap containing key value pairs for container metadata
+   * @throws AzureBlobFileSystemException
+   */
+  public Hashtable<String, String> getContainerMetadata(TracingContext tracingContext)
+      throws AzureBlobFileSystemException {
+    try (AbfsPerfInfo perfInfo = startTracking("getContainerMetadata", "getContainerMetadata")) {
+      LOG.debug("getContainerMetadata for filesystem: {}", client.getFileSystem());
+
+      final AbfsRestOperation op = client.getContainerMetadata(tracingContext);
+      perfInfo.registerResult(op.getResult()).registerSuccess(true);
+
+      final Hashtable<String, String> metadata = parseResponseHeadersToHashTable(op.getResult());
```

Comment:
common code in getFilesystemProperties as well:
```
  final AbfsRestOperation op = client.getContainerMetadata(tracingContext);
 final Hashtable<String, String> metadata = parseResponseHeadersToHashTable(op.getResult());
```

should we make seperate method and use that both places.

Reply:
Same as setFileSystemProperties.
We will still end up with one line repeated..